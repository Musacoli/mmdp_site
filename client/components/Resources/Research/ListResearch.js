import React from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'semantic-ui-react';
import BlueCard from '../../common/Card/BlueCard';
import ConfirmModal from '../../common/Modal/ConfirmModal';
import InvalidPage from '../../common/InvalidPage';
import {
  ucFirstLetter,
  ARCHIVE_ACTION,
  UNARCHIVE_ACTION,
} from '../../../utils/helper';
import Search from '../../common/Search';
import AddNewButton from '../../common/AddNewButton';

const ListResearch = ({
  results,
  isOpen,
  modalAction,
  addResearchUrl,
  instanceName,
  showModal,
  handleModalToggle,
  onConfirm,
  handleSearch,
  handleSearchChange,
  loading,
}) => {
  const generateEditLink = (id) => `/resources/research/edit/${id}`;
  const researchResults = results.results;
  let modalInfo;
  if (modalAction === 'delete') {
    modalInfo = {
      header: 'Delete research',
      content: 'Are you sure you want to delete research?',
    };
  } else if (
    modalAction === ARCHIVE_ACTION ||
    modalAction === UNARCHIVE_ACTION
  ) {
    modalInfo = {
      header: `${ucFirstLetter(modalAction)} research`,
      content: `Are you sure you want to ${modalAction} research?`,
    };
  }
  return (
    <>
      <Grid>
        <Grid.Column width={12}>
          <Search
            onSearch={handleSearch}
            onChange={handleSearchChange}
            className="research-search"
            placeholder="Search research"
          />
        </Grid.Column>
        <Grid.Column width={3}>
          <AddNewButton
            url={addResearchUrl}
            text="Research"
            className="common__button bg-ugly-blue"
          />
        </Grid.Column>
      </Grid>
      <Grid relaxed>
        {researchResults && researchResults.length
          ? researchResults.map(({ _id: id, title, archived }) => {
              const archiveAction = archived
                ? UNARCHIVE_ACTION
                : ARCHIVE_ACTION;
              return (
                <Grid.Column key={id} mobile={8} tablet={8} computer={4}>
                  <BlueCard
                    title={title}
                    archived={archived}
                    onArchive={(event) =>
                      showModal(event, { id, modalAction: archiveAction })
                    }
                    onDelete={(event) =>
                      showModal(event, { id, modalAction: 'delete' })
                    }
                    editLink={generateEditLink(id)}
                  />
                </Grid.Column>
              );
            })
          : !loading && (
              <Grid.Row className="ui loading center aligned animated fadeIn">
                <InvalidPage
                  pathLabel={`Add a ${instanceName}`}
                  errorMessage={`No ${instanceName} to display`}
                  errorDescription={`Please add ${instanceName}`}
                  path={addResearchUrl || '/'}
                />
              </Grid.Row>
            )}
        <ConfirmModal
          isOpen={isOpen}
          handleModalToggle={handleModalToggle}
          onConfirm={onConfirm}
          modalInfo={modalInfo}
        />
      </Grid>
    </>
  );
};

ListResearch.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  modalAction: PropTypes.string,
  showModal: PropTypes.func.isRequired,
  handleModalToggle: PropTypes.func.isRequired,
  results: PropTypes.shape({}).isRequired,
  onConfirm: PropTypes.func.isRequired,
  handleSearchChange: PropTypes.func.isRequired,
  handleSearch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

export default ListResearch;
