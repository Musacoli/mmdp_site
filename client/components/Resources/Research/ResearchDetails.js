import React from 'react';
import { List } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

export const ResearchCard = ({ title, Archived, _id, onDelete, onArchive }) => {
  return (
    <div className="research__details">
      <p id="research__content">{title}</p>
      <List horizontal>
        <List.Item>
          <Link to={`/resources/research/edit/${_id}`} id="research__edit__btn">
            Edit
          </Link>
        </List.Item>
        <List.Item
          as="a"
          id="research__archive__btn"
          onClick={() => {
            const data = Archived ? { Archived: false } : { Archived: true };
            onArchive(_id, data);
          }}
        >
          {Archived ? 'UnArchive' : 'Archive'}
        </List.Item>
        <List.Item
          as="a"
          id="research__delete__btn"
          onClick={() => {
            onDelete(_id);
          }}
        >
          Delete
        </List.Item>
      </List>
    </div>
  );
};

export default ResearchCard;

ResearchCard.propTypes = {
  title: PropTypes.string.isRequired,
  Archived: PropTypes.bool.isRequired,
  _id: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onArchive: PropTypes.func.isRequired,
};
