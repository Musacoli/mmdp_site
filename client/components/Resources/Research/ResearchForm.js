import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import FileInput from '../../common/Form/FileInput';

const ResearchForm = ({ fileName, onChange, onSubmit, loading }) => (
  <Form onSubmit={onSubmit} loading={loading} className="main-content">
    <Form.Group widths="equal">
      <Form.Field>
        <FileInput
          name="researchFile"
          label="Upload file"
          placeholder="Select a file"
          value={fileName || ''}
          className="test"
          id="researchFile"
          onChange={onChange}
        />
      </Form.Field>
      <Form.Input
        name="title"
        className="form__text-input"
        fluid
        label="Research title"
        type="text"
        placeholder="Research carried out last year"
        onChange={onChange}
      />
    </Form.Group>
    <Form.Button type="submit" className="btn__upload u-margin-top-medium">
      Upload Research
    </Form.Button>
  </Form>
);

ResearchForm.propTypes = {
  loading: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  fileName: PropTypes.string.isRequired,
};

export default ResearchForm;
