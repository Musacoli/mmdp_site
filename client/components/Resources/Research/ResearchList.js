import React from 'react';
import { Grid } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import Details from './ResearchDetails';

export const ResearchList = ({ results }) => {
  const researchResults = results.results;
  let researchDetails;
  if (researchResults) {
    researchDetails = researchResults.map((x) => {
      return (
        <Details
          title={x.title}
          key={x._id}
          _id={x._id}
          Archived={x.Archived}
        />
      );
    });
  }
  return (
    <div>
      <Grid className="Research__list">{researchDetails}</Grid>
    </div>
  );
};

export default ResearchList;

ResearchList.propTypes = {
  // results: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  results: PropTypes.shape({}).isRequired,
};
