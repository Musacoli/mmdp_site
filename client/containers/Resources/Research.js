import React, { Component } from 'react';
import { connect } from 'react-redux';
import ResearchView from '../../components/Resources/Research/ResearchView';
import { addResearch } from '../../store/actions/resources/research';
import SideBarNav from '../Sidebar';

export class Research extends Component {
  state = {
    title: '',
    reportFile: '',
    fileName: '',
  };

  onChange = (e) => {
    if (e.target.name === 'reportFile' && e.target.files.length) {
      const file = e.target.files[0];
      this.setState({
        fileName: file.name || '',
        reportFile: file,
      });
    } else {
      this.setState({ [e.target.name]: e.target.value });
    }
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    const details = this.state;
    const formData = new FormData();
    Object.keys(details).forEach((key) => {
      formData.append(key, details[key]);
    });
    this.props.addResearch(formData);
  };

  render() {
    return (
      <div className="research__container">
        <div>
          <SideBarNav />
        </div>
        <div id="research__view">
          <ResearchView
            onChange={this.onChange}
            onSubmit={this.onFormSubmit}
            fileName={this.state.fileName}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  Research: state.response,
});

const mapDispatchToProps = {
  addResearch,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Research);
