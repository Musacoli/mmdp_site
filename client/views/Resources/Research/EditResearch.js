import React from 'react';
import Templates from '../../Templates';
import EditResearchContainer from '../../../containers/Resources/Research/EditResearch';

const EditResearchView = ({ ...props }) => (
  <Templates {...props} title="Edit Research">
    <EditResearchContainer {...props} />
  </Templates>
);

export default EditResearchView;
