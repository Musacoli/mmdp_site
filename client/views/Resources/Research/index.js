import React from 'react';
import Template from '../../Templates';
import AddResearch from '../../../containers/Resources/Research';

const ResearchView = ({ ...props }) => (
  <Template {...props} title="Add research">
    <AddResearch {...props} />
  </Template>
);

export default ResearchView;
