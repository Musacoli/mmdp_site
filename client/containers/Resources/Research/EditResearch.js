import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ResearchForm from '../../../components/Resources/Research/ResearchForm';
import EmptyView from '../../../components/common/InvalidPage';
import {
  getResearchRequest,
  updateInputData,
  updateResearchRequest,
} from '../../../store/actions/resources/research';

export class EditResearch extends Component {
  componentWillMount() {
    const { getResearch, match } = this.props;
    const { id } = match.params;
    getResearch(id);
  }

  onChange = (e) => {
    const { InputData, updateData } = this.props;
    if (e.target.name === 'researchFile' && e.target.files.length) {
      const file = e.target.files[0];
      InputData({
        ...updateData,
        researchFile: file,
        Filename: file.name,
      });
    } else {
      InputData({ ...updateData, [e.target.name]: e.target.value });
    }
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    const { updateResearch, updateData, match } = this.props;
    const { id } = match.params;

    if (updateData.researchFile && updateData.researchFile.url) {
      delete updateData.researchFile;
    }

    const formData = new FormData();
    Object.keys(updateData).forEach((key) => {
      formData.append(key, updateData[key]);
    });
    updateResearch({ formData, id });
  };

  render() {
    const { researchItem, loading, updateData, error } = this.props;

    return (
      <div>
        {researchItem.data ? (
          <ResearchForm
            onChange={this.onChange}
            onSubmit={this.onFormSubmit}
            fileName={updateData.Filename || ''}
            title={researchItem.data.title}
            loading={loading}
          />
        ) : (
          <div>
            {error.data ? (
              <EmptyView
                errorMessage="Research Not Found"
                errrorDescription="The Research you are trying to Edit Does not Exist"
                path="/resources/research/all"
                pathLabel="View Research Items"
              />
            ) : (
              <span />
            )}
          </div>
        )}
      </div>
    );
  }
}

EditResearch.propTypes = {
  getResearch: PropTypes.func.isRequired,
  updateResearch: PropTypes.func.isRequired,
  InputData: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
  updateData: PropTypes.shape({}).isRequired,
  researchItem: PropTypes.shape({}).isRequired,
  match: PropTypes.shape({}).isRequired,
  error: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state) => ({
  loading: state.editResearch.loading,
  researchItem: state.editResearch.researchItem,
  updateData: state.editResearch.updateInput,
  error: state.editResearch.error,
});

const mapDispatchToProps = {
  getResearch: getResearchRequest,
  InputData: updateInputData,
  updateResearch: updateResearchRequest,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditResearch);
