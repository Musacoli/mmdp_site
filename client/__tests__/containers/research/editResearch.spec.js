import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReactRouterEnzymeContext from 'react-router-enzyme-context';
import { EditResearch } from '../../../containers/Resources/Research/EditResearch';

describe('<Edit Research Container /> ', () => {
  const match = {
    params: { id: 1 },
  };

  const updateData = {
    title: 'Welcome To Research',
    researchFile: {
      bucket: 'mmdp-img-assets',
      etag: '9bd46a8adedb4a78a165c4ad6c302a25',
      filename: 'VHl2-3o5I-DdnL-Q',
      mimetype: 'application/pdf',
      path: '/assets/documents',
      size: 519957,
      url:
        'https://s3.amazonaws.com/mmdp-img-assets/assets/documents/VHl2-3o5I-DdnL-Q',
    },
  };

  const wrapper = shallow(
    <EditResearch
      match={match}
      getResearch={jest.fn}
      InputData={jest.fn}
      updateData={updateData}
      researchItem={{}}
      updateResearch={jest.fn}
      error={{}}
    />,
    new ReactRouterEnzymeContext(),
  );

  it('renders Edit Research conatiner without crashing', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Except Wrapper To Have These Functions', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'title',
        value: 'This is My Title',
        checked: true,
        files: ['File'],
      },
    };

    const fileEvent = {
      preventDefault: jest.fn(),
      target: {
        name: 'researchFile',
        files: [
          {
            name: 'file.pdf',
          },
        ],
      },
    };
    wrapper.instance().onFormSubmit(event);
    wrapper.instance().onChange(event);
    wrapper.instance().onChange(fileEvent);
  });
});
