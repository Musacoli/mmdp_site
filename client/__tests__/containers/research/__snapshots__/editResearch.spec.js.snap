// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Edit Research Container />  renders Edit Research conatiner without crashing 1`] = `
<div>
  <div>
    <span />
  </div>
</div>
`;
